tests/integration/test_rerelate.py

```diff
@@ -0,0 +1,166 @@
+tester_app_name = "prometheus-tester"
+tester_resources = {
+    "prometheus-tester-image": oci_image(
+        "./tests/integration/prometheus-tester/metadata.yaml",
+        "prometheus-tester-image",
+    )
+}
+
+
+async def test_setup_env(ops_test: OpsTest):
+    await ops_test.model.set_config({"logging-config": "<root>=WARNING; unit=DEBUG"})
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest, prometheus_charm, prometheus_tester_charm):
+    """Build the charm-under-test and deploy it together with related charms."""
+    await asyncio.gather(
+        ops_test.model.deploy(
+            prometheus_charm,
+            resources=resources,
+            application_name=app_name,
+            num_units=2,
+            trust=True,
+        ),
+        ops_test.model.deploy(
+            prometheus_tester_charm,
+            resources=tester_resources,
+            application_name=tester_app_name,
+        ),
+        ops_test.model.deploy(
+            "ch:alertmanager-k8s",
+            application_name="alertmanager",
+            channel="edge",
+        ),
+        ops_test.model.deploy(
+            "ch:grafana-k8s",
+            application_name="grafana",
+            channel="edge",
+        ),
+        ops_test.model.deploy(
+            "ch:grafana-agent-k8s",
+            application_name="grafana-agent",
+            channel="edge",
+        ),
+        # TODO: add traefik after fetch-lib (there are breaking changes)
+    )
+
+    await asyncio.gather(
+        ops_test.model.add_relation(app_name, tester_app_name),
+        ops_test.model.add_relation(app_name, "alertmanager"),
+        ops_test.model.add_relation(app_name, "grafana:grafana-source"),
+        ops_test.model.add_relation(app_name, "grafana-agent:send-remote-write"),
+    )
+    await ops_test.model.wait_for_idle(status="active", timeout=600)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_relation(ops_test: OpsTest):
+    await asyncio.gather(
+        ops_test.model.applications[app_name].remove_relation("metrics-endpoint", tester_app_name),
+        ops_test.model.applications[app_name].remove_relation("alertmanager", "alertmanager"),
+        ops_test.model.applications[app_name].remove_relation("grafana-source", "grafana"),
+        ops_test.model.applications[app_name].remove_relation(
+            "receive-remote-write", "grafana-agent"
+        ),
+    )
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=600)
+
+
+@pytest.mark.abort_on_fail
+async def test_rerelate(ops_test: OpsTest):
+    await asyncio.gather(
+        ops_test.model.add_relation(app_name, tester_app_name),
+        ops_test.model.add_relation(app_name, "alertmanager"),
+        ops_test.model.add_relation(app_name, "grafana:grafana-source"),
+        ops_test.model.add_relation(app_name, "grafana-agent:send-remote-write"),
+    )
+    await ops_test.model.wait_for_idle(status="active", timeout=600)
```

Comment:
I'm not sure why there are different tests for removing the relation, re-relating, then removing. We'd get a traceback no matter where, and an integration test with only two calls seems over-normalized.

Reply:
As an aside, using the "actual" app names rather than "implicit all" maybe somewhat more reliable.

If it's not specified, it's implicitly `self.applications -> property(self._live_entity_map("applications"))`. The property will be hit every time, and the "live entity map" can change in the meantime, because it may be [updated by a watcher asynchronously](https://github.com/juju/python-libjuju/blob/bd9490d1c0abebf0fe324ecdfcb43e3b30ac993f/juju/model.py#L1044), and properties [assigned to variables](https://github.com/juju/python-libjuju/blob/bd9490d1c0abebf0fe324ecdfcb43e3b30ac993f/juju/model.py#L2446) are called every time. Like [in a `while True:` loop](https://github.com/juju/python-libjuju/blob/bd9490d1c0abebf0fe324ecdfcb43e3b30ac993f/juju/model.py#L2472)

Relying on the implicit here without any synchronization primitives or libjuju actually snapshotting the state of the value is almost asking for `apps` to be different every time it wakes up from sleep. We should be explicit.